`clang` does not detect GNU extensions in Haiku

Using clang 12 on Haiku `hrev57210`, with this source file:

```c
#include <math.h>
#include <stdint.h>

int main()
{
	return (int)(intptr_t)(void*)&sincos;
}
```

Compilation using `gcc test.c -o test -D_GNU_SOURCE` works normally.

However, compilation using `clang` fails:

```
~/Desktop> clang test.c -o test -D_GNU_SOURCE
test.c:6:32: error: use of undeclared identifier 'sincos'
        return (int)(intptr_t)(void*)&sincos;
                                      ^
1 error generated.
```

> make sure /system/develop/headers/gnu is in the include path before /system/develop/headers/posix

The error still occurs.
```
/SystemRoot/home/trung> BEINCLUDES=/boot/system/develop/headers/gnu:$BEINCLUDES clang test.c -o test -D_GNU_
SOURCE
test.c:6:32: error: use of undeclared identifier 'sincos'
        return (int)(intptr_t)(void*)&sincos;
                                      ^
1 error generated.
```

This is a full verbose log of a `clang` invocation:

```
/SystemRoot/home/trung> BEINCLUDES=/boot/system/develop/headers/gnu:$BEINCLUDES clang -v test.c -o test -D_G
NU_SOURCE
clang version 12.0.1
Target: x86_64-unknown-haiku
Thread model: posix
InstalledDir: /bin
 "/boot/system/bin/clang-12" -cc1 -triple x86_64-unknown-haiku -emit-obj -mrelax-all --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -fhalf-no-semantic-interposition -mframe-pointer=all -fmath-errno -fno-rounding-math -mconstructor-aliases -munwind-tables -target-cpu x86-64 -tune-cpu generic -fno-split-dwarf-inlining -debugger-tuning=gdb -v -resource-dir /boot/system/lib/clang/12.0.1 -D _GNU_SOURCE -fdebug-compilation-dir /SystemRoot/home/trung -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -faddrsig -o /tmp/test-2b9889.o -x c test.c
clang -cc1 version 12.0.1 based upon LLVM 12.0.1 default target x86_64-unknown-haiku
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/boot/system/non-packaged/develop/headers"
ignoring nonexistent directory "/boot/system/develop/headers/os/opengl"
ignoring nonexistent directory "/boot/system/develop/headers/os/be_apps/Deskbar"
ignoring nonexistent directory "/boot/system/develop/headers/3rdparty"
ignoring nonexistent directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /boot/system/lib/clang/12.0.1/include
 /boot/system/develop/headers/os
 /boot/system/develop/headers/os/app
 /boot/system/develop/headers/os/arch
 /boot/system/develop/headers/os/device
 /boot/system/develop/headers/os/drivers
 /boot/system/develop/headers/os/game
 /boot/system/develop/headers/os/interface
 /boot/system/develop/headers/os/kernel
 /boot/system/develop/headers/os/locale
 /boot/system/develop/headers/os/mail
 /boot/system/develop/headers/os/media
 /boot/system/develop/headers/os/midi
 /boot/system/develop/headers/os/midi2
 /boot/system/develop/headers/os/net
 /boot/system/develop/headers/os/storage
 /boot/system/develop/headers/os/support
 /boot/system/develop/headers/os/translation
 /boot/system/develop/headers/os/add-ons/graphics
 /boot/system/develop/headers/os/add-ons/input_server
 /boot/system/develop/headers/os/add-ons/mail_daemon
 /boot/system/develop/headers/os/add-ons/registrar
 /boot/system/develop/headers/os/add-ons/screen_saver
 /boot/system/develop/headers/os/add-ons/tracker
 /boot/system/develop/headers/os/be_apps/NetPositive
 /boot/system/develop/headers/os/be_apps/Tracker
 /boot/system/develop/headers/bsd
 /boot/system/develop/headers/glibc
 /boot/system/develop/headers/posix
 /boot/system/develop/headers
End of search list.
test.c:6:32: error: use of undeclared identifier 'sincos'
        return (int)(intptr_t)(void*)&sincos;
                                      ^
1 error generated.
/SystemRoot/home/trung>
```

Seems like it does not recognize the `gnu` headers at all.

--- a/sys-devel/llvm/llvm12-12.0.1.recipe
+++ b/sys-devel/llvm/llvm12-12.0.1.recipe
@@ -31,7 +31,7 @@ other than the ones listed above.
 HOMEPAGE="https://www.llvm.org/"
 COPYRIGHT="2003-2019 University of Illinois at Urbana-Champaign"
 LICENSE="Apache v2 with LLVM Exception"
-REVISION="7"
+REVISION="8"
 SOURCE_URI="https://github.com/llvm/llvm-project/releases/download/llvmorg-$portVersion/llvm-$portVersion.src.tar.xz"
 CHECKSUM_SHA256="7d9a8405f557cefc5a21bf5672af73903b64749d9bc3a50322239f56f34ffddf"
 SOURCE_DIR="llvm-$portVersion.src"

--- a/sys-devel/llvm/llvm12-12.0.1.recipe
+++ b/sys-devel/llvm/llvm12-12.0.1.recipe
@@ -456,7 +456,7 @@ PROVIDES_clang_analysis="
 	"
 REQUIRES_clang_analysis="
 	llvm12${secondaryArchSuffix}_clang == $portVersion base
-	cmd:python
+	cmd:python3
 	"
 CONFLICTS_clang_analysis="
 	llvm${secondaryArchSuffix}_clang_analysis
@@ -503,13 +503,14 @@ REQUIRES_libs="
 	lib:libz$secondaryArchSuffix
 	"
 
-PYTHON3_VERSION="3.9"
-PROVIDES_python3="
-	llvm12${secondaryArchSuffix}_python3 = $portVersion
+PYTHON_PACKAGES="python310"
+PYTHON3_VERSION="3.10"
+PROVIDES_python310="
+	llvm12${secondaryArchSuffix}_$PYTHON_PACKAGES = $portVersion
 	cmd:lit
 	"
-REQUIRES_python3="
-	setuptools_python${PYTHON3_VERSION//.}
+REQUIRES_python310="
+	setuptools_$PYTHON_PACKAGES
 	cmd:python$PYTHON3_VERSION
 	"
 
@@ -557,7 +558,7 @@ REQUIRES_libunwind_devel="
 BUILD_REQUIRES="
 	haiku${secondaryArchSuffix}_devel
 	devel:libz$secondaryArchSuffix
-	setuptools_python39
+	setuptools_$PYTHON_PACKAGES
 	"
 BUILD_PREREQUIRES="
 	cmd:cmake
@@ -633,7 +634,7 @@ INSTALL()
 	ninja install
 
 	# GENERIC: all python_setuptools-based installs need this
-	local pythonPackageName="${portName}_python3-$portFullVersion"
+	local pythonPackageName="${portName}_python310-$portFullVersion"
 	local packageLinksDir=$(dirname $portPackageLinksDir)
 	python=$packageLinksDir/${pythonPackageName}/cmd~python$PYTHON3_VERSION/bin/python$PYTHON3_VERSION
 	pythonVersion=$($python --version 2>&1 | sed 's/Python //' | head -c3)
@@ -806,8 +807,8 @@ INSTALL()
 		$libDir/libLTO* \
 		$libDir/libRemarks*
 
-	# python3 package
-	packageEntries python3 \
+	# python310 package
+	packageEntries python310 \
 		$binDir/lit \
 		$prefix/lib/python*
 

--- a/sys-devel/llvm/patches/clang-12.0.1.patchset
+++ b/sys-devel/llvm/patches/clang-12.0.1.patchset
@@ -194,3 +194,28 @@ index 50a20c2..ac65381 100644
 -- 
 2.37.3
 
+
+From 089a07be1f393fc06a8b67f015e40a718ba550dc Mon Sep 17 00:00:00 2001
+From: Trung Nguyen <57174311+trungnt2910@users.noreply.github.com>
+Date: Mon, 14 Aug 2023 14:58:56 +1000
+Subject: [PATCH] Haiku: Add missing GNU headers include path
+
+---
+ lib/Frontend/InitHeaderSearch.cpp | 1 +
+ 1 file changed, 1 insertion(+)
+
+diff --git a/lib/Frontend/InitHeaderSearch.cpp b/lib/Frontend/InitHeaderSearch.cpp
+index b8ecee03..2d999898 100644
+--- a/lib/Frontend/InitHeaderSearch.cpp
++++ b/lib/Frontend/InitHeaderSearch.cpp
+@@ -331,6 +331,7 @@ void InitHeaderSearch::AddDefaultCIncludePaths(const llvm::Triple &triple,
+     AddPath("/boot/system/develop/headers/3rdparty", System, false);
+     AddPath("/boot/system/develop/headers/bsd", System, false);
+     AddPath("/boot/system/develop/headers/glibc", System, false);
++    AddPath("/boot/system/develop/headers/gnu", System, false);
+     AddPath("/boot/system/develop/headers/posix", System, false);
+ #ifdef HAIKU_HYBRID_SECONDARY
+     AddPath("/boot/system/develop/headers/" HAIKU_HYBRID_SECONDARY, System, false);
+-- 
+2.39.2
+

